.. This document is autogenerated by CommandList.reStructuredTextHelp().
   Do not edit it here. Edit source code then run tests in CommandListTest.updateReStructuredFile() to recreate this file.

Command reference
=================

This is the documentation for the indvidual commands. The help documented here can be invoked also at the command prompt with `command -h`, for example to get the help for `ylim`::

    ylim -h

Parameters in square brakets are optional and the default argument is indicated by the `=` sign. The syntax `...` indicate that the argument can be repeated multiple times. For example::

    ylim min max [track_regex = .*]...

Means that `ylim` takes two mandatory arguments, `min` and `max`. The optional argument, `track_regex`, defaults to `.*` and can be repated multiple times.


Navigation
----------

goto
++++

:code:`goto chrom:[from]-[to]`

Go to region `chrom:from-to` or to `chrom:from` or to the start of `chrom`.  The character ':' is a shortcut for `goto`. Examples::

    goto chr8:1-1000  ## Go to interval 1-1000 on chr8
    goto chr8:10      ## Go to position 10 on chr8
    goto chr8         ## Go to start of chr8

Or the same with::

    :chr8:1-1000 
    :chr8:10 
    :chr8


INT
+++

:code:`INT [INT]`

Go to position `INT` or to region `INT INT` on current chromosome.  Allowed is the hyphenated format  separating the two positions. If a list of integers is given, the first and last are taken as *from* and *to*. This is handy to copy and paste intervals from the ruler above the prompt. 
Examples::

    10                   -> Will jump to position 10 
    10 1000              -> Go to region 10-1000 
    10-1000              -> Same as above
    10 250 500 750 1000  -> Same as above again



plus +
++++++

:code:`+ INT [k|m]`

Move forward by `INT` bases. Suffixes k (kilo) and M (mega) are expanded to x1000 and x1,000,000. Examples::

    +2m
    +10k
    +10.5k

 

minus -
+++++++

:code:`- INT [k|m]`

Move backwards by INT bases. Suffixes k (kilo) and M (mega) are expanded to x1000 and x1,000,000.
Examples::

    -100
    -10k
    -10.5m

 

f - forward
+++++++++++

:code:`f [NUM=0.1]`

Move forward NUM times the size of the current window, 1/10 by default. 

b - backward
++++++++++++

:code:`b [NUM=0.1]`

Move backward NUM times the size of the current window, 1/10 by default 

ff
++

:code:`ff`

Move forward by 1/2 of a window. A shortcut for `f 0.5` 

bb
++

:code:`bb`

Move backward by 1/2 of a window. A shortcut for `b 0.5` 

zi
++

:code:`zi [INT = 1]`

Zoom in INT times. Each zoom halves the window size.  To zoom quickly use INT= 5 or 10 e.g. `zi 10`

zo
++

:code:`zo [INT = 1]`

Zoom out INT times. Each zoom doubles the window size.  To zoom quickly use INT= 5 or 10 e.g. `zo 10`

extend
++++++

:code:`extend [mid|window] [INT left] [INT right]`

Extend the current window by `INT` bases left and right.
 
* :code:`window` (default): Extend the current window left and right by `INT` bases

* :code:`mid`: The new window is given by the midpoint of the current window plus and minus `INT` bases left and right.

If only one INT is given it is applied to both left and right. Negative INTs will shrink instead of extend the window.

l - left
++++++++

:code:`l`

Go to the Left half of the current window.  Alternate the left and right command to quickly focus on a point of interest. 

r - right
+++++++++

:code:`r`

Go to the Right half of the current window.  Alternate the left and right command to quickly focus on a point of interest. 

p
+

:code:`p`

Go to the previous visited position.  Similar to the back and forward arrows of an Internet browser.

n
+

:code:`n`

Go to the next visited position.  Similar to the back and forward arrows of an Internet browser.

next
++++

:code:`next [-start] [track]`

Move to the next feature on `track` on current chromosome.  `next` centers the window on the found feature and zooms out. This is useful for quickly browsing through annotation files of genes or ChIP-Seq peaks in combination with read coverage tracks (bigwig, tdf, etc.).

* :code:`-start`: Sets the window right at the start of the feature, without centering and zooming out.

The `next` command does exactly that, it moves to the next feature. If there are no more features after the current position it doesn't rewind to the beginning (use `1` for that) and it doesn't move to another chromosome, use `goto chrom` for that.
 
If `track` is omitted, the first annotation track is used.

Find
----

find
++++

:code:`find [-all] regex [track]`

Find the first record in `track` containing `regex`. The search for `regex` starts from the *end* of the current window (so the current window is not searched) and moves forward on the current chromosome. At the end  of the current chromosome move to the next chromosomes and then restart at  the start of the initial one. The search stops at the first match found. If `track` is omitted the first interval track found is searched.
The :code:`-all` flag will return the region containing **all** the regex matches.
Examples::

    find -all ACTB genes.gtf -> Find all the matches of ACTB
    find 'ACTB gene'         -> Find the first match of 'ACTB gene'

Use single quotes to define patterns containing spaces.

seqRegex
++++++++

:code:`seqRegex [-iupac] [-c] [regex]`

Find regex in reference sequence and show matches as an additional track.  Options:

* :code:`regex`: Regex to search. If missing the seq regex track is removed.

* :code:`-iupac`: Enable the interpretation of the IUPAC ambiguity code. NB: This option simply converts IUPAC chracters to the corresponding regex.

* :code:`-c`: Enable case-sensitive matching. Default is to ignore case.

Examples::

    seqRegex ACTG        -> Case insensitive, actg matched
    seqRegex -c ACTG     -> Case sensitive, will not match actg
    seqRegex -iupac ARYG -> Interpret (converts) R as [AG] and Y as [CT]
    seqRegex             -> Disable regex matching track

To save matches to file, see the `print` command. This command is ignored if the reference fasta sequence is missing.

bookmark
++++++++

:code:`bookmark [name] | [-rm] | [-print] | [> [file]]`

Creates a track to save positions of interest. Without arguments, add the current position to the bookmark track. Options:

* :code:`name`: give this name to the new bookmark.

* :code:`-rm`: remove the bookmark matching *exactly* the current position.

* :code:`-print`: prints to screen the list of current bookmarks.

* :code:`>`: saves the bookmark to file.

Examples::

    bookmark              -> Add the current position to bookmarks.
    bookmark myGene       -> Add the current position with name myGene
    bookmark -rm          -> Remove the bookmark exactly in this position
    bookmark > books.txt  -> Save to file books.txt
    bookmark -print       -> Show table of bookmarks



Display
-------

grep
++++

:code:`grep [-i = .*] [-e = ''] [track_regex = .*]...`

Similar to grep command, filter for features including or excluding patterns. Options:

* :code:`-i regex`:  Show features matching this regex.

* :code:`-e regex`: Exclude features matching this regex.

* :code:`track_regex`: Apply to tracks matched by `track_regex`.

Regex `-i` and `-e` are applied to the raw lines as read from source file. This command is useful to filter the annotation in GTF or BED files, for example::

    grep -i RNA -e mRNA gtf gff

Will show the rows containing 'RNA' but will hide those containing 'mRNA', applies to tracks whose name matches 'gtf' or 'gff'.
With no arguments reset to default: :code:`grep -i .* -e ^$ .*` which means show everything, hide nothing, apply to all tracks.

squash
++++++

:code:`squash [track_regex = .*]...`

Toggle the squashing of features with the same coordinates.  If set, features with the same start, end, and strand are squashed in a single one. The displayed feature is the first one found in the group of features with the same coordinates. Useful to compact GTF where e.g. CDS and exons have the same coordinates. Applies only to annotation tracks captured by track_regex

merge
+++++

:code:`merge [track_regex = .*]...`

Toggle the merging of overlapping features. If set, features with overalapping coordinates are merged in a single one. Merged features will not have strand and name information. Note that merging is done without considering strand information. Applies only to annotation tracks captured by the list of track_regex

gap
+++

:code:`gap [track_regex = .*]...`

Toggle the switch to add a gap between features. Default is true. If gap is set, as per default, features which on screen do not have at least one space separating are moved to different lines so that it is clear where is the end of one feature  and the start of the next one. If gap is unset such features might appear as a single, continuous one instaed. 
Example with gap set::

    ||||||
          ||||||

With gap unset these two features look like::

    ||||||||||||

Gap unset is preferable when the interest is in knowing which regions are covered since it gives a more compact view and the distiction betwen adjacent features is not important.

gffNameAttr
+++++++++++

:code:`gffNameAttr [attribute_name = NULL] [track_regex = .*]...`

For GTF/GFF tracks, choose the attribute to get the feature name from. Use attribute NULL to reset to default choice of attribute. Applies to all GFF/GTF tracks captured by the list of `track_regex`. Example, given the gtf feature::

    chr1 . CDS  10 99 . + 2 gene_id "PTGFRN"; transcript_id "NM_020440";

Use gene_name as feature name or transcript_id::

    gffNameAttr gene_name genes.gtf .*gff
    PTGFRN_CCCCCCCCC
    
    gffNameAttr transcript_id genes.gtf .*gff
    NM_020440_CCCCCC



trackHeight
+++++++++++

:code:`trackHeight INT [track_regex = .*]...`

Set track height to INT lines of text for all tracks matching regexes.  Setting height to zero hides the track and skips the processing altogether. This is useful to speed up the browsing when large bam files are present. Use infoTrack to see which tracks are hidden. Example::

    trackHeight 5 aln.*bam gtf`


ylim
++++

:code:`ylim <NUM|min|na> <NUM|min|na> [track_regex = .*]...`

Set the y-axis limit for all tracks matched by regexes. The first two arguments set the min and max limits. The 3rd argument is a list of regexes to capture the tracks to reset. Argument min and max can be:

* :code:`NUM`: Numeric, fix the limits exactly to these values

* :code:`na`: Scale tracks to their individual min and/or max

* :code:`min` and :code:`max`: Set to the min and max of **all** tracks.

This command applies only to tracks displaying quantitative data on y-axis (e.g. bigwig, tdf), the other tracks are unaffected. Examples::

    ylim 0 50      -> Set min= 0 and max= 50 in all tracks.
    ylim 0 na      -> Set min to 0 and autoscale the max. Apply to all tracks
    ylim na na tdf -> Autoscale min and max. Apply to all tracks matching 'tdf'
    ylim min max   -> Set to the min and max of all tracks



colorTrack
++++++++++

:code:`colorTrack color [track_regex = .*]...`

Set colour for tracks matched by regex.  Available colours: red, green, yellow, blue, magenta, cyan, grey, light_red, light_green, light_yellow, light_blue, light_magenta, light_cyan, light_grey, white, black, default. The 'default' colour reset to the system default colour. Colouring is rendered with ANSI codes 8/16. Example::

    colorTrack light_blue ts.*gtf ts.*bam



hideTitle
+++++++++

:code:`hideTitle [track_regex = .*]...`

Toggle the display of the title line matched by track_regex. Use /hide_all/ and /show_all/ to hide all tracks or show all tracks instead of toggling their mode. Examples::

    hideTitle            -> Toggle all tracks, same as hideTitle .*
    hideTitle bam bed    -> Toggle all tracks matched by 'bam' or 'bed'
    hideTitle /hide_all/ -> Hide all tracks


editNames
+++++++++

:code:`editNames -t <pattern> <replacement> [track_re=.*]...`

Edit track names by substituting regex pattern with replacement. Pattern and replacement are required arguments, the default regex for track is '.*' (i.e. all tracks).
The :code:`-t` (test) flag shows what renaming would be done without actually editing the names.
Use "" (empty double quotes) to replace pattern with nothing. Examples: Given track names 'fk123_hela.bam#1' and 'fk123_hela.bed#2'::

    editNames fk123_ ""    - > hela.bam#1, hela.bed#2
    editNames fk123_ "" bam -> hela.bam#1, fk123_hela.bed#2
    editNames _ ' '           -> fk123 hela.bam#1,  fk123 hela.bed#2
    editNames ^.*# cells      -> cells#1, cells#2
    editNames ^ xx_           -> xx_fk123_hela.bam#1, xx_fk123_hela.bed#2 (add prefix)


dataCol
+++++++

:code:`dataCol [index = 4] [track_regex = .*]...`

Select data column for bedgraph tracks containing regex.  index: 1-based column index. This command applies only to tracks of type bedgraph.
 For example, use column 5 on tracks containing #1 and #3::
 
    dataCol 5 #1 #3



print
+++++

:code:`print [-full] [-off] [track_regex = .*]... [>|>> file]`

Toggle the printing of lines for the tracks matched by `track_regex`.  Useful to show exactly what features are present in the current window. Features are filtered in/out according to the :code:`grep` command. Options:

* :code:`track_regex`: Toggle printing of the tracks matched by one or more of these regexes.

* :code:`-full`: Wrap lines longer than the screen width. Default is to clip them.

* :code:`-off`: Turn off printing for *all* tracks, regardless of their current mode. The list of regexes is effectively ignored and set to '.*'.

* :code:`>` and :code:`>>`: Send the output to `file` instead of to screen. `>` overwrites existing file, `>>` appends. Redirecting to file is probably not useful if more than track is selected as files will overwrite each other or be appended. The %r variable is expanded to the current genomic coordinates.

Examples::
    print                        -> Print all tracks, same as `print .*`
    print -off                   -> Turn off printing for all tracks
    print genes.bed >> genes.txt -> Append features in track(s) 'genes.bed' to file

Currently `print` applies only to annotation tracks, other tracks are unaffected.

Alignments
----------

rpm
+++

:code:`rpm [track_regex = .*]`

Toggle read coverage from raw count to reads per million. Applys to BAM and TDF files. 

samtools
++++++++

:code:`samtools [-f INT=0] [-F INT=4] [-q INT=0] [track_re = .*] ...`

Apply samtools filters to alignment tracks captured by the list of track regexes. As *samtools view*, this command filters alignment records on the basis of the given flags:

* :code:`-F`: Filter out flags with these bits set. NB: 4 is always set.

* :code:`-f`: Require alignment to have these bits sets.

* :code:`-q`: Require alignments to have MAPQ >= than this.

Examples::

    samtools -q 10           -> Set mapq for all tracks. -f and -F reset to default
    samtools -F 1024 foo bar -> Set -F for all track containing re foo or bar
    samtools                 -> Reset all to default.


BSseq
+++++

:code:`BSseq [track_regex = .*]...`

Toggle bisulfite mode for read tracks matched by regex. In bisulfite mode, the characters M and m mark methylated bases (i.e. unconverted C to T) and U and u are used for unmethylated bases (i.e. C converted to T). Upper case is used for reads on  forward strand, small case for reverse. Ignored without reference fasta sequence.

General
-------

setGenome
+++++++++

:code:`setGenome fasta|bam|genome`

Set genome and reference sequence. The genome, i.e. the list of contig and names and sizes, can be extracted from the indexed fasta reference, from a bam file or from a genome identifier (e.g. hg19). If a fasta file is used also the reference sequence becomes available.

showGenome
++++++++++

:code:`showGenome`

Print the genome dictionary with a representation of chromosome sizes.  Example output::

    showGenome
    chrM  16571
	   chr1  249250621 ||||||||||||||||||||||||||||||
    chr2  243199373 |||||||||||||||||||||||||||||
    ...
    chr21 48129895  ||||||
    chr22 51304566  ||||||
    chrX  155270560 |||||||||||||||||||
    chrY  59373566  |||||||



infoTracks
++++++++++

:code:`infoTracks`

Print the name of the current tracks along with file name and format.  Hidden tracks are marked by an asterisk.

addTracks
+++++++++

:code:`addTracks [file or URL]...`

Add tracks from local or remote files. 
Examples::

    addTracks peaks.bed gene.gtf
    addTracks http://remote/host/peaks.bed


dropTracks
++++++++++

:code:`dropTracks [-t] track_regex [track_regex]...`

Drop tracks matching any of the listed regexes. The :code:`-t` (test) flag only shows what tarcks would be removed without actually removing them.
Examples::

    dropTracks bam


orderTracks
+++++++++++

:code:`orderTracks [track_regex]...`

Reorder tracks according to the list of regexes or sort by name. Not all the tracks need to be listed, the missing ones follow the listed ones in unchanged order. Without arguments sort track by tag name.
For example, given the track list: `[hela.bam#1, hela.bed#2, hek.bam#3, hek.bed#4]`::

    orderTracks #2 #1   -> [hela.bed#2, hela.bam#1, hek.bam#3, hek.bed#4]
    orderTracks bam bed -> [hela.bam#1, hek.bam#3, hela.bed#2, hek.bed#4]
    orderTracks         -> name sort [hela.bam#1, hela.bed#2, hek.bam#3, hek.bed#4]


history
+++++++

:code:`history`

Show the list of visited positions. 

cmdHistory
++++++++++

:code:`cmdHistory`

Show the list of executed commands. 

save
++++

:code:`save [filename = chrom_start_end.txt']`

Save current screenshot to file in either text or png format. Default filename is generated from the current coordinates and the default format is txt. With filename .png save as png using current coordinates as filename. Use extension .png to save as png format. Colours are stripped from text output but they retainedin the png. The string `%r` in the file name is replaced with the current coordinates.Examples::

    save mygene.txt    -> Save to mygene.txt as text
    save               -> Save to chrom_start-end.txt as text
    save .png          -> Save to chrom_start-end.png as png
    save mygene.%r.png -> Save to mygene.chr1_100-200.png as png



sessionSave
+++++++++++

:code:`sessionSave filename`

Experimental: Save the current settings to file suitable to be reloaded by ASCIIGenome. `sessionSave` writes to file a set of commands to reproduce the current settings: tracks, colors, heights etc. It's not meant to be a perfect replica, rather it's a shortcut to avoid re-typing commands. Example::

    sessionSave session.txt

Quit session and reload with::

    ASCIIGenome -x session.txt


q
+

:code:`q`

Quit 

h
+

:code:`h -h`

h and -h show this help. For help on specific commands use `command -h`, for example :code:`ylim -h` 


